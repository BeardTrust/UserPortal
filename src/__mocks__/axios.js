export default {
    post: jest.fn().mockResolvedValue({
        response: {
            data: {},
            headers: {}
        }
    }),
    put: jest.fn().mockResolvedValue({
        response: {
            data: {}
        }
    }),
    get: jest.fn().mockResolvedValue({
        data: {
            user: {
                userId: 'abc-123-xyz-789',
                email: 'a@b.com',
                username: '*',
                password: '*',
                firstName: '*',
                lastName: '*',
                phone: '3',
                dateOfBirth: '1234-01-01'
            }
        }
    })
}