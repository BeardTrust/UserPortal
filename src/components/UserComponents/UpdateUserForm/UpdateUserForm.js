import axios from "axios"
import { useRef } from "react"
import AuthContext from "../../../store/auth-context"
import { useContext, useState } from "react"
import { Form, FormGroup, FormLabel, FormControl, ButtonGroup, Button } from "react-bootstrap"


function UpdateUserForm({ user }) {
    const authContext = useContext(AuthContext);
    const [errorMessage, setErrorMessage] = useState();

    const emailRef = useRef();
    const firstNameRef = useRef();
    const lastNameRef = useRef();
    const passwordRef = useRef();
    const phoneRef = useRef();
    const usernameRef = useRef();
    const dateOfBirthRef = useRef();
    const token = authContext.token;
    const userId = authContext.userId;


    const url = 'http://localhost:9001/users/' + userId

    function submitUpdate(event) {
        if (!event === null) {
            event.preventDefault();
        }

        let enteredEmail
        let enteredUsername
        let enteredPassword
        let enteredFirstName
        let enteredLastName
        let enteredPhone
        let enteredDateOfBirth
        if (emailRef.current.value !== null) {enteredEmail = emailRef.current.value}
        if (firstNameRef.current.value !== null) {enteredFirstName = firstNameRef.current.value;}
        if (lastNameRef.current.value  !== null) {enteredLastName = lastNameRef.current.value;}
        if (passwordRef.current.value  !== null) {enteredPassword = passwordRef.current.value;}
        if (phoneRef.current.value  !== null) {enteredPhone = phoneRef.current.value;}
        if (usernameRef.current.value  !== null) {enteredUsername = usernameRef.current.value;}
        if (dateOfBirthRef.current.value  !== null) {enteredDateOfBirth = dateOfBirthRef.current.value;}


        if (emailRef.current.value === '') {
            enteredEmail = user.email
        } if (firstNameRef.current.value === '') {
            enteredFirstName = user.firstName;
        } if (lastNameRef.current.value === '') {
            enteredLastName = user.lastName;
        } if (passwordRef.current.value === '') {
            enteredPassword = user.password;
        } if (phoneRef.current.value === '') {
            enteredPhone = user.phone
        } if (usernameRef.current.value === '') {
            enteredUsername = user.username
        } if (dateOfBirthRef.current.value === '') {
            enteredDateOfBirth = user.dateOfBirth
        }
        const registrationData = {
            email: enteredEmail,
            firstName: enteredFirstName,
            lastName: enteredLastName,
            password: enteredPassword,
            phone: enteredPhone,
            username: enteredUsername,
            dateOfBirth: enteredDateOfBirth
        }
    

        axios(
            url,
            {
                method: 'PUT',
                data: JSON.stringify(registrationData),
                headers: {
                    'Authorization': token,
                    'Content-Type': 'application/json'
                }
            })
            .then((response) => {
                return response.data;
            }).then((data) => {
                console.log('UPDATE SUCCESSFUL');
                console.log(data);
            }).catch((e) => {
                setErrorMessage(e);
                console.log('UPDATE ERROR')
                console.log(token)
                if (e.response.status === 409) {
                    console.log('Error 409 (Bad Request) detected. This is likely a duplicate user.')//need an error page responding with duplicae warning
                }
                if (e.response.status === 503) {
                    console.log('Error 503 (Service Unavailablet) detected. The server is likely down.')//need an error page responding with duplicae warning
                } 
                else {
                    console.log('Unexpected error: ', errorMessage + ' please contact an admin.')
                }
            });

    }


    return (
        <section>
            <div className={'container vertical-center'}>
                <Form className={'offset-4 col-3'}>
                <h3>Want to update anything?</h3>
                    <FormGroup>
                        <FormLabel htmlFor={'username'} className={'col-form-label'}>Update Username</FormLabel>
                        <FormControl type={'text'} id={'username'} ref={usernameRef} />
                    </FormGroup>
                    <FormGroup>
                        <FormLabel htmlFor={'password'}>Update Password</FormLabel>
                        <FormControl type={'password'} id={'password'} ref={passwordRef} />
                    </FormGroup>
                    <FormGroup>
                        <FormLabel htmlFor={'firstName'}>Update First Name</FormLabel>
                        <FormControl type={'text'} id={'firstName'} ref={firstNameRef} />
                    </FormGroup>
                    <FormGroup>
                        <FormLabel htmlFor={'lastName'}>Update Last Name</FormLabel>
                        <FormControl type={'text'} id={'lastName'} ref={lastNameRef} />
                    </FormGroup>
                    <FormGroup>
                        <FormLabel htmlFor={'email'}>Update Email Address</FormLabel>
                        <FormControl type={'email'} id={'email'} ref={emailRef} email={'true'} />
                    </FormGroup>
                    <FormGroup>
                        <FormLabel htmlFor={'phone'}>Update Phone Number</FormLabel>
                        <FormControl type={'text'} id={'phone'} ref={phoneRef}  />
                    </FormGroup>
                    <FormGroup>
                        <FormLabel htmlFor={'dateOfBirth'}>Update Date of Birth</FormLabel>
                        <FormControl type={'text'} id={'dateOfbirth'} ref={dateOfBirthRef}  />
                    </FormGroup>
                    <ButtonGroup>
                        <Button type={'submit'} className={'btn btn-primary mt-3'} onClick={submitUpdate}>Submit Update</Button>
                    </ButtonGroup>
                </Form>
            </div>
        </section>
    )
}
export default UpdateUserForm