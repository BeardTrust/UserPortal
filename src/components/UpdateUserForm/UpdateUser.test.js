import axios from 'axios'
import TestRenderer from 'react-test-renderer'
import UpdateUserForm from './UpdateUserForm'
import {act, cleanup, render, screen} from '@testing-library/react';
import userEvent from '@testing-library/user-event';

/**
 * The tests for the ViewUserForm component.
 *
 * @author Nathanael Grier <Nathanael.Grier@Smoothstack.com>
 */
jest.mock('axios');
describe('UpdateUserForm PUT Request and formatting: ', () => {
    /**
     * This method tests to see if the Update PUT Request is submitted via axios.
     */
    it('Puts data in the backend', async () => {
        const data = {
            data: {
                user: {
                    userId: 'cf8bd72a-4b4c-42f1-8dc3-06cc2dc2cb8a',
                    email: 'example@email.com',
                    username: 'user',
                    password: 'pass',
                    firstName: 'Fname',
                    lastName: 'Lname',
                    phone: '3303046134',
                    dateOfBirth: '1997-09-05'
                }
            }
        };
        const user = axios.put.mockImplementationOnce(() => Promise.resolve(data));
        const updateUserForm = render(<UpdateUserForm user={user} url={'/test'}/>);
        const submitButton = screen.getByText('Submit Update');
        // userEvent.click(submitButton);
        // await act(() => user);
    });

    /**
     * This method tests to see if the axios.post request logs the returned user id.
     */
    it('Expects the screen to make sure the snapshot matches', async () => {
        const tr = TestRenderer.create(<UpdateUserForm/>);
        expect(tr.toJSON()).toMatchSnapshot();
    });
});