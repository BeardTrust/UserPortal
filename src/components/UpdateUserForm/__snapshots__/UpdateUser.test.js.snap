// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateUserForm PUT Request and formatting:  Expects the screen to make sure the snapshot matches 1`] = `
<section>
  <div
    className="container vertical-center"
  >
    <form
      className="offset-4 col-3"
    >
      <h3>
        Want to update anything?
      </h3>
      <div>
        <label
          className="col-form-label form-label"
          htmlFor="username"
        >
          Update Username
        </label>
        <input
          className="form-control"
          id="username"
          type="text"
        />
      </div>
      <div>
        <label
          className="form-label"
          htmlFor="password"
        >
          Update Password
        </label>
        <input
          className="form-control"
          id="password"
          type="password"
        />
      </div>
      <div>
        <label
          className="form-label"
          htmlFor="firstName"
        >
          Update First Name
        </label>
        <input
          className="form-control"
          id="firstName"
          type="text"
        />
      </div>
      <div>
        <label
          className="form-label"
          htmlFor="lastName"
        >
          Update Last Name
        </label>
        <input
          className="form-control"
          id="lastName"
          type="text"
        />
      </div>
      <div>
        <label
          className="form-label"
          htmlFor="email"
        >
          Update Email Address
        </label>
        <input
          className="form-control"
          email="true"
          id="email"
          type="email"
        />
      </div>
      <div>
        <label
          className="form-label"
          htmlFor="phone"
        >
          Update Phone Number
        </label>
        <input
          className="form-control"
          id="phone"
          type="text"
        />
      </div>
      <div>
        <label
          className="form-label"
          htmlFor="dateOfBirth"
        >
          Update Date of Birth
        </label>
        <input
          className="form-control"
          id="dateOfbirth"
          type="text"
        />
      </div>
      <div
        className="btn-group"
        role="group"
      >
        <button
          className="btn btn-primary mt-3 btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="submit"
        >
          Submit Update
        </button>
      </div>
    </form>
  </div>
</section>
`;
